backend/settings.js

```diff
@@ -59,7 +59,10 @@ const DEFAULT_SETTINGS = {
      * Generally set through the ENV variable:
      * PLOTLY_CONNECTOR_IS_RUNNING_INSIDE_ON_PREM
      */
-    IS_RUNNING_INSIDE_ON_PREM: false
+    IS_RUNNING_INSIDE_ON_PREM: false,
+
+    // Whether it is running in Circle CI environment
+    IS_RUNNING_ON_CIRCLE_CI: false
```

Comment:
No need for this. Please, use https://circleci.com/docs/2.0/env-vars/#circleci-environment-variable-descriptions

Reply:
Circle env vars cannot be used with `getSetting` because the environment variables used in falcon are of the `PLOTLY_CONNECTOR_*` format.

Reply:
There is no need to use `getSetting`, use `process.env.CIRCLECI` instead.

Reply:
I think we should be consistent in our usage. We **dont** use `process.env` anywhere in code for accessing environment variables.

Going forward if you intend to use `process.env` for accessing environment variables, please open a separate issue where we can get all environment variables use `process.env.*`.

Reply:
> I think we should be consistent in our usage. We dont use process.env anywhere in code for accessing environment variables.

That's not true.
`process.env` is used by both the backend and the frontend.
Note that it's also used by `webpack.config.*`.

`getSetting` is only used to access properties in `~/.plotly/connector/settings.yaml` and ENV variables starting with `PLOTLY_CONNECTOR`.

For simplicity's sake, please, implement this change as requested.

Reply:
No, We don't. We only use it in `webpack.config.*` because we cannot access `getSettings` there. Here is a grep excluding `webpack.*` and compiled bundles:

```
 ✝  ~/plotly/falcon-sql-client   gke-deployments±  ack 'process.env' -r .
app/store/configureStore.js
1:if (process.env.NODE_ENV === 'production') {

test/backend/settings.spec.js
16:        delete process.env.PLOTLY_CONNECTOR_STORAGE_PATH;
21:        delete process.env.PLOTLY_CONNECTOR_STORAGE_PATH;
32:        process.env.PLOTLY_CONNECTOR_STORAGE_PATH = __dirname;
74:        process.env.PLOTLY_CONNECTOR_ACCESS_TOKEN_EXPIRY = String(now);

.eslintrc
55:        "no-process-env": [0],

webpack.config.headless.js
27:            'process.env.NODE_ENV': JSON.stringify('production')

backend/menus.js
86:            submenu: (process.env.NODE_ENV === 'development') ? [{
153:            submenu: (process.env.NODE_ENV === 'development') ? [{

backend/settings.js
158: * Load settings from process.env prefixed with `PLOTLY_CONNECTOR_`
165:    } else if (has(`PLOTLY_CONNECTOR_${settingName}`, process.env)) {
166:        const envObject = process.env[`PLOTLY_CONNECTOR_${settingName}`];
```

Note that we only check for `NODE_ENV`. The only other place is `settings.spec.js` which is used to modify how `getSetting` behaves in tests.

Reply:
> No, We don't.

The results you're quoting show that `process.env` is also used in `app/store/configureStore.js`, `backend/menus.js`. 


It's possible to use github to see that it's also used in `backend/main.development.js` and `backend/headless.development.js`. See https://github.com/plotly/falcon-sql-client/search?utf8=%E2%9C%93&q=process.env&type= :

![image](https://user-images.githubusercontent.com/6199391/34780284-a26f42f2-f61a-11e7-8bc7-09852f7a7985.png)

![image](https://user-images.githubusercontent.com/6199391/34780329-c72b5338-f61a-11e7-91a7-1de4bbab43b7.png)

![image](https://user-images.githubusercontent.com/6199391/34780366-df8ff424-f61a-11e7-896a-23c99573c33e.png)

![image](https://user-images.githubusercontent.com/6199391/34780451-19fb6d00-f61b-11e7-9f6a-dff9de440caa.png)

![image](https://user-images.githubusercontent.com/6199391/34780472-2899d158-f61b-11e7-8c78-01428c4aecc4.png)

Reply:
Anyway, as I explained above `getSetting` isn't meant to access the ENV variable `CIRCLECI`.

The ENV variable `CIRCLECI` is set automatically by CircleCI and we can access it simply by reading `process.env.CIRCLECI`.

What you're proposing here is that:
1. we add the seeting `IS_RUNNING_ON_CIRCLE_CI` to `backend/settings.js`
2. we set `PLOTLY_CONNECTOR_IS_RUNNING_ON_CIRCLE_CI` in `circle.yml`
3. and we read the value by calling `getSetting('IS_RUNNING_ON_CIRCLE_CI')`

Please, for the sake of simplicity, implement the requested change; i.e. just use `process.env.CIRCLECI` to read the value.

Reply:
Using `getSetting` looks like a clearer and more consistent solution than sprinkling `process.env` reads throughout the code. It's best to centralize configuration handling in one place with one mechanism, and `getSetting` does that.

I support reading `process.env.CIRCLECI` in order to determine if the setting should be set, but it sounds like there are good arguments for doing it the way @tarzzz has implemented it now, namely that it's consistent with the way the other settings are handled.

Reply:
@scjody I strongly disagree. `getSetting` isn't meant to access properties like 
 `process.env.CIRCLECI` (as I explained above).

Reply:
@n-riesco I don't see any technical reason why `getSetting` couldn't be enhanced to deal with `process.env.CIRCLECI` as well as environment variables beginning with `PLOTLY_CONNECTOR_`. I suggested it because if you don't want to add another environment variable then that provides a way forward while still using `getSetting`.

`getSetting` is the way we handle settings in the backend of this app. If you don't want to use it with `process.env.CIRCLECI` then let's just add another environment variable as @tarzzz is proposing.

Sprinkling `process.env.CIRCLECI` checks throughout the code will result in something that's significantly less maintainable since you're adding a second configuration mechanism. (There is one `process.env` user in the backend as you pointed out, but it predates the existence of `getSetting` and should probably be changed for consistency. But that's a separate concern.)

Reply:
@scjody 

> I don't see any technical reason why getSetting couldn't be enhanced to deal with process.env.CIRCLECI as well as environment variables beginning with PLOTLY_CONNECTOR_.

Technically it's possible; from a design point of view, it is undesirable. I've posted a number of reasons and none of them hasn't been refuted yet (I'll list them again at this end of this comment).

> getSetting is the way we handle settings in the backend of this app. If you don't want to use it with `process.env.CIRCLECI` then let's just add another environment variable as @tarzzz is proposing.

The backend doesn't need to know about `process.env.CIRCLECI`, only the specs do.

> Sprinkling process.env.CIRCLECI checks throughout the code will result in something that's significantly less maintainable since you're adding a second configuration mechanism. 

As I showed in my previous comments, the code is already sprinkled with `process.env`. I agree with you that `process.env` should be removed from the code (but this will have to wait until I have a chance to work again on the `webpack.config.*.js`).

Anyway, this argument doesn't apply to `process.env.CIRCLECI`, because it'd be used only inside the tests.

---

To sum up, here's the list of arguments against introducing the setting `IS_RUNNING_ON_CIRCLE_CI`:
1. The simplest solution (that requires no changes in the code) is to use `process.env.CIRCLECI`. Introducing `IS_RUNNING_ON_CIRCLE_CI` requires changes in `backend/settings.js`.
2. `CIRCLECI` is set automatically by CircleCI  (`PLOTLY_CONNECTOR_IS_RUNNING_ON_CIRCLE_CI` would require to change `circle.yml`)
3. Falcon doesn't need to know about `process.env.CIRCLECI` (only the tests do).